### ISSUE TITLE
Implement new Easy/Normal/Hard/Brutal AI with meaningful differences

### ISSUE THREAD

**Issue by pchote** [OpenRA/OpenRA#16126]
Our current AI definitions in the default mods arent that great.  General feedback from casual players shows that they are simultaneously far too difficult for new players, and trivially easy for experienced players.  The different bot types that only differ in unit compositions also goes against common expectations.

We've grown options over the years that let us implement meaningful difficulty levels, but there has so far been little interest in enabling them.

I would like to see us get rid of our current confusing set of bots, and standardise on Easy, Normal, Hard, Brutal options across all mods. Each of these types can then pick a random build composition based on our current bot choices by using `GrantRandomCondition` to select between `UnitBuilderBotModule`s.

**Normal AI**
Basically the same as our current bots, but with:
* Small reduction to support power targeting efficiency (#16123)
* Small random delay (0 - 5 second) when queuing structures (#16124)
* Possibly a smaller random delay (0 - 2 second) when queuing units

**Easy AI**
Basically the same as Normal AI, but with:
* Further reduced support power targeting efficiency
* Further increased delay for structure and unit queuing
* Smaller `SquadSize` so that attacks are easier to counter
* A relatively low cap on refinery and harvester count to limit economy

**Hard AI**
Basically the same as Normal AI, but with:
* Optimal (but non-cheaty) support power targeting
* No delays when queuing units or structures
* Enabled AutoTarget priorities so rocket soldiers prefer targeting vehicles over infantry, flak trucks prefer aircraft and then infantry over tanks, and so on

**Brutal AI**
Basically the same as Hard AI, but cheats:
* 125% (or 150%?) income/production speed/health/damage bonuses
* Support power targeting ignores visibility (#16123)

**Comment by ABrandau** [comment]
From experience with SP and CD, I wanted to add that an AI with infinite money (Cash Trickler) is a monster to face, tho this is seen even more in mods with multi queue, this can be challenging but not that entertaining,

**Comment by ABrandau** [comment]
Also, maybe you could consider adding the option to force the AI to go for a build path, its the AI randomness what makes its difficulty so unpredictable.

**Comment by NotoriousPyro** [comment]
Please rename 'Brutal AI' to 'Cheat AI' for clarity.

The AI is exceptionally easy because it does not try to expand enough. It destroys an enemy and then doesn't even expand to new areas of tiberium. It plays with one con yard and I've only ever seen it create several con yards a few times, and that was after I'd popped it's con yard to see what it would do.

There are huge areas of blue tiberium which it practically ignores for long amounts of time. I've also spectated a match where the AI was building a replacement refinery, at the same time as a base defense after all its refineries had been destroyed. Instead of cancelling the base defense, the AI sat practically the whole game (15 minutes), about half a second from completing that refinery - all it had to do was cancel the base defense production.

It would help if the AI was more aggressive with its expansion as opposed to just 'letting it cheat' - that way it wouldn't get into these money-less positions.

**Comment by NotoriousPyro** [comment]
![image](https://user-images.githubusercontent.com/706132/73611845-2429d900-45de-11ea-9088-9273eaf55a8f.png)

Screenshot for clarity.

**Comment by vext01** [comment]
I have fond memories of playing RA skirmishes as a kid, so I installed OpenRA to re-live some of that.

But my goodness, skirmishes are way too hard for a casual player.

**Comment by pchote** [comment]
The "Brutal AI" can be removed in favor of an explicit handicap system (either based on #17937 or something like it), with options to make the bots (or the player) easier (handicap > 0) or harder (handicap < 0).

**Comment by rask** [comment]
I'm with @vext01 here. I remember playing RA95 as a kid for fun against AI, but playing against OpenRA AI is just stressful and not fun for folks like me who like their strategy games with a slower pace. Glad to hear this is on the design table!

**Comment by DarkDuneSand** [comment]
Specifically made an account for this. I played all Dune games when they came out and played them a lot. I am a player that just like to enjoy it. It is a moment of relaxing. But not this. Tried 4 skirmishes now and for me they are all impossible to play (dead after 10-15 minutes). And no, I do not want to get better at this game, I want to enjoy it. This is just too stressful and unplayable for me, no fun at all. Waaay to hard. To be clear, i am not really a gamer. I did not like to play any other RTS games but the old Dune2, 2k and Emperor where really a lot of fun for me. But OpenRA is just way harder then the original. And I bet a lot of people just drop this game if it only can run at this difficulty level.

How do I Enjoy this game? Building a big base and defend it for hours and hours. Make it bigger and bigger. Maybe boring for others, that is what I liked back then, and I still do now.

**Comment by vext01** [comment]
This is off-topic, but wanted to say: if you enjoy Dune 2, there is the
Dune Legacy remake which is much easier than OpenRA. I also played a lot of
Dune 2 in the DOS era :)

http://dunelegacy.sourceforge.net/website/

**Comment by pchote** [comment]
For now, you might want to try [this test build](https://github.com/pchote/OpenRA/releases/tag/playtest-20201213-hc), which adds the handicap feature from #18971 to the current playtest-20201213. Setting a handicap on the bots will slow them down and make it easier to defend when they do attack.

**Comment by mooleshacat** [comment]
Yeah the game is impossible regardless of the settings you put it on. You get bumrushed no matter what settings. You can't even get a base started before they already attacking you. Then the AI you are supposedly allied with stands around while the enemy attack your base.

Tried the multiplayer but humans rush you and all are better than you and the game ends (at least for you) in a few seconds/minutes.

I went singleplayer/skirmish to try and control the difficulty and make a lengthy game where you can build a base and some tanks before being attacked. Problem is the AI rushes you right off the start no matter what usually with success.

It would be nice to have an easy/medium/hard/insane setting.

**Comment by Bakaba** [comment]
Why are ALL free RTS bots suffer from the exact same issue? Of course you don't code an AI that cheats at you when you can code an AI smart enough to kill you without cheating! Instead of buffing up the unfair advantage of AIs, why not make them smarter first, and add an optional "resources buff" where you can choose up what to give them. Also, explains how the AI work in the AI selection screen and get rid of that "Entire map-sentient" things.

All that above are common sense when you played at least 1 game of any RTS. Guess you were born master gamer of RTS and started developing RTS in your mother's belly or something to not know this.

**Comment by WyattBest** [comment]
I know this is GitHub, not a forum for reminiscing, but these comments are hilarious.

> Yeah the game is impossible regardless of the settings you put it on. You get bumrushed no matter what settings. You can't even get a base started before they already attacking you. Then the AI you are supposedly allied with stands around while the enemy attack your base.

This sounds just like my dad when I played the originals with him as a kid. Now I'm playing OpenRA with my young son and having a great time! We can beat an AI together, but he gets steamrolled by himself.

What if there were a level below Easy AI (Tutorial AI? Passive AI?) that does not attack until one at least one of its units gets within range of an enemy or a timer expires? In other words, the AI builds blissfully until engaged, then starts behaving normally. Many beginner/child/casual players want to build a lot before doing any fighting. Totally unrealistic, but it's what people want.

**Comment by tbertels** [comment]
Fog of war (not unexplored map) should be off (at least by default) in easy and normal mode.
See also #14912.
